### ISSUE TITLE
`range(-1)` iterates forever

### ISSUE THREAD

**Issue by mtfishman** [ITensor/ITensor#221]
The following iterates forever:

```c++
for( auto i : range(-1) ) println("i = ",i);
```

Maybe the `range(...)` constructor should throw an error when a negative range is input, or set the end to 0?

I came across this trying to track down a subtle bug when converting a TagSet to a string, where I wanted to iterate over all tags except the last one using range (so `for( auto i : range(size(ts)-1) )`), but then when there were no tags it turned out to iterate forever. Probably there is a better way to write the function, but it is definitely not expected behavior.

**Comment by emstoudenmire** [comment]
Yes, I think the solution is just to check for invalid input to the range function. By the way, I was thinking we should ask Nils if he would just put range (an improved version of it) into his itertools and we can bring it into ITensor that way.

**Comment by mtfishman** [comment]
Yeah that would be good.
